### ISSUE TITLE
FOV names and array names are hard-coded for the HCS dataset reader

### ISSUE THREAD

**Issue by ziw-liu** [ome/ome-zarr-py#248]
The [v0.4 spec states](https://ngff.openmicroscopy.org/0.4/index.html#image-layout):

```
# The name of the array is arbitrary with the ordering defined by
# by the "multiscales" metadata, but is often a sequence starting at 0.
```

And ome-zarr-py works with arbitrary array names in the single-FOV images layout, but not in an HCS plate layout. Only integer names starting with "0" is recognized.

It should instead be discovered from metadata (`multiscales.datasets[0].path`).

Also, FOV names not starting with "0" is not supported. The cause is likely this line hard-coding 'first field':

https://github.com/ome/ome-zarr-py/blob/c1302e05998dfe2faf94b0f958c92888681f5ffa/ome_zarr/reader.py#L493

It should instead be discovered from metadata (`well.images[0].path`). [Spec](https://ngff.openmicroscopy.org/0.4/index.html#well-md):

```
The well dictionary MUST contain an images key whose value MUST be a list of JSON objects specifying all fields of views for a given well. Each image object MUST contain a path key whose value MUST be a string specifying the path to the field of view. The path MUST contain only alphanumeric characters, MUST be case-sensitive, and MUST NOT be a duplicate of any other path in the images list. 
```

Edit: separate description for array and position names.

**Comment by ziw-liu** [comment]
According to https://github.com/ome/ome-zarr-py/issues/200#issuecomment-1131586455, one can view all the FOVs by loading individual wells. However this only works for incremental integer FOV path names, not any 'case sensitive, alphanumeric' path names defined in the specification. Possible cause:

https://github.com/ome/ome-zarr-py/blob/c1302e05998dfe2faf94b0f958c92888681f5ffa/ome_zarr/reader.py#L427-L431

**Comment by will-moore** [comment]
Thanks for the feedback.
You're right that there is an assumption made about the path to first image in a Well.
It would be a pretty easy fix to load the Well images for a Well and read that data. It's a bit of a bigger change to read that data for ALL Wells in a plate if you can't assume that all Wells have the same path to first image.

This is a similar type of change to https://github.com/ome/ome-zarr-py/pull/241 where it becomes slower to load plate data if you don't make assumptions about all the Wells being the same.

Reading Well images may not have such a big impact as the changes on that PR, but it raises similar questions:
For plates where all Wells have the same path to first image, we would prefer NOT to have to load the Images for every Well to check.

@ziw-liu  In your data, does each Well have a different path to the first image?
If so, then I guess we first need to test and see how much slower it is to read that for each image.

**Comment by ziw-liu** [comment]
@will-moore Thanks for the explanation.

I agree that hardcoding is the fastest way to get the images. As for the dataset I have in mind, all the wells would have the same path (in fact `'0'`) to the first image. I was trying to load synthetic data (generated by testing a custom writer) with the viewer, and found that writing with different FOV names that should be valid according to the specification is only supported by the ome-zarr-py reader when loading a single well.

The array name finding is of a bit more real-world concern. In our workflow it is common to transform the raw data into other arrays with incompatible data types and shapes, and it would be a cleaner data management practice if these can all be held in the same Zarr group with their names indicating their nature (e.g. `'raw'`, `'projection'`). It is possible to rename the arrays we want to load with ome-zarr-py and napari-ome-zarr to `'0'` so that they are recognized, but it would be nice if the NGFF specification is fully implemented here, and data can be accessed with Zarr-style name keys when not using ome-zarr-py.

As for the performance concern, I don't have a intuition for cloud object storage, but for a storage server in the local netowrk, increased I/O operations should only be linear to the image count to load the `.zattrs` files. But of course only testing can give a real answer.

**Comment by will-moore** [comment]
Thanks for the info. This makes good sense.
Do you have a 'multiscales' object for e.g. `raw` and `projection`, each with a pyramid of 'datasets'?
Currently, only the first multiscales is object is read by ome-zarr-py.

**Comment by ziw-liu** [comment]
> Do you have a 'multiscales' object for e.g. `raw` and `projection`, each with a pyramid of 'datasets'?

I do not think we will have multiple pyramids for a single FOV, and I agree that for visualization only loading the first pyramid makes sense. Are you suggesting that incompatible arrays can be stored in separate `multiscales` objects?

**Comment by will-moore** [comment]
The spec allows multiple `multiscales` images https://ngff.openmicroscopy.org/latest/#multiscale-md and doesn't say anything about the relationship between them.
However, currently all NGFF tools assume only 1 multiscales image per NGFF file and ignore any others: (see 'multiple multiscales' at https://ome.github.io/ome-ngff-tools/)

**Comment by aeisenbarth** [comment]
I fell into the same problem. Well image path and field index seem to be used a bit interchangeably in the code, but this is not reflected in any document, especially the spec. For my use case, I'd prefer to have human-readable image names, because these are displayed, e.g. in Napari.

If the reason are performance concerns about accessing image paths when reading the plate, then maybe the paths should not (only) live at the well level, or the spec could be modified not to allow non-index image names.

Since ome-zarr-py already has assumptions (not fully implemented spec), it's a matter of weighing up the probability that a dataset has well image path != field index versus the probability that the first well's image paths are not the same for all wells. I think the second is less likely and a fix would increase the amount of supported datasets.

**Comment by ziw-liu** [comment]
As of v0.10.3 I'm still seeing this bug. Consider this test case which passes:

```py
def test_fov_name_extern(tmp_path):
    import numpy as np
    import zarr

    from ome_zarr.io import parse_url
    from ome_zarr.writer import (
        write_image,
        write_plate_metadata,
        write_well_metadata,
    )
    from ome_zarr.reader import Reader

    path = tmp_path / "test_ngff_plate.zarr"
    images = (
        ("A", "1", "0"),
        ("B", "1", "0"),
        ("H", "12", "0"),
    )
    row_names = ["A", "B", "H"]
    col_names = ["1", "12"]
    well_paths = [f"{row}/{col}" for row, col, _ in images]
    shape = (4, 128, 128)

    store = parse_url(path, mode="w").store
    root = zarr.group(store=store)
    write_plate_metadata(root, row_names, col_names, well_paths)
    for wi, wp in enumerate(well_paths):
        row, col = wp.split("/")
        row_group = root.require_group(row)
        well_group = row_group.require_group(col)
        field = images[wi][2]
        write_well_metadata(well_group, field)
        image_group = well_group.require_group(field)
        write_image(
            image=np.ones(shape),
            group=image_group,
            axes="zyx",
            storage_options=dict(chunks=shape),
        )
    list(Reader(parse_url(path))())
```

With a simple change:

```diff
    images = (
        ("A", "1", "0"),
        ("B", "1", "0"),
-       ("H", "12", "0"),
+       ("H", "12", "1"),
    )
```

The test fails because ome-zarr-py is still trying to look for "H/12/0":

```
self = <zarr.storage.FSStore object at 0x10c676c90>, key = 'H/12/0/.zgroup'
value = b'{\n    "zarr_format": 2\n}'

    def __setitem__(self, key, value):
        if self.mode == "r":
>           raise ReadOnlyError()
E           zarr.errors.ReadOnlyError: object is read-only

.../site-packages/zarr/storage.py:1463: ReadOnlyError
```

I think this issue needs to be re-opened, since this behavior is not only a spec conformance bug but also a round trip failure.

**Comment by will-moore** [comment]
Agreed that the current behaviour doesn't follow the spec exactly, because of the trade-off with performance.

We could load the Well fields for every Well, but that could be a lot of extra calls which would be expected to slow down the loading of a plate. How much slower I don't know, but we probably don't want to do it for ALL plates, since many won't need it.

One option that occurs to me is that for every Well, we could try to get the image data using the `first_field_path` (as we do now), but if that fails, then instead of returning zeros, we could load the Well fields and pick the first at

https://github.com/aeisenbarth/ome-zarr-py/blob/9d247f4eaaa2e97e3bb77cc735add8285627c64f/ome_zarr/reader.py#L549

This would have no impact on performance when every Well uses the same `first_field_path`, but would handle the odd Well where the field path(s) are different.

For a plate where most Wells use a different field path from the first Well, this try/except strategy could be slower than simply looking up the field paths for every Well first. If we know that the try/except is going to fail for most Wells then it's best not to waste time with it. But in the case where we only need to handle *exceptional* Wells, then this feels like a nice solution.

Thoughts @ziw-liu @jluethi ?

**Comment by jluethi** [comment]
Thanks for pinging me on this discussion @will-moore . I think it's another part of highlighting some current limits of the plate spec for efficient visualization. Good to think of ways to work around this for now & see whether in the collection RFC space, we can then come up with future plate collections that would avoid some of the issues we observed here (e.g., the question that I ask myself here: is it really beneficially to have separate well metadata or should there be a single, slightly larger plate metadata object that would contain the list of images per well?)

Also big-picture: While I'm not a fan of just considering the images in the well as FOVs (see https://github.com/ome/ngff/pull/137), we also often have different images in a well (e.g. raw vs. processed or multiple rounds of acquisition of the same well). In these cases, it would be interesting to be able to load a defined image for each well instead of always loading image 0. We currently solve this for ourselves in the napari plugin context by offering another plugin that can inspect the OME-Zarr & load arbitrary images/labels/tables for wells one-by-one (see https://github.com/fractal-napari-plugins-collection/napari-ome-zarr-navigator).

On the specific issue here: I'm not familiar with any `first_field_path` metadata in the OME-Zarr itself. Is the proposal here that this is an ome-zarr-py default and ome-zarr-py would expose a `first_field_path` variable that the user could change? And in case `first_field_path` isn't found in a well, the well metadata is loaded to determine the first image in the order? 
Or we don't expose it as an option, but keep it hard-coded at "0" and just run into the try/except loop if the "0" doesn't exist?
Or is there actually a `first_field_path` metadata somewhere in plate OME-Zarrs itself?

**Comment by ziw-liu** [comment]
@will-moore I cannot evaluate the performance impact of fixing this bug without knowing the underlying file system and the connection type that you are thinking about. I personally work with on-premise HPC systems and parsing 10^2 JSON files (i.e. for a 96-well plate) has not been an issue in my own implementation. Of course things can be very different in a high latency or low resource setting. As long as the behavior is correct (as in spec-conformant), I am supportive.
